src/apps/chifra/internal/abis/handle_addresses.go

```diff
@@ -0,0 +1,113 @@
+package abisPkg
+
+import (
+	"context"
+	"os"
+	"sort"
+
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/abi"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/account"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/logger"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/output"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/types"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+func (opts *AbisOptions) HandleAddresses() (err error) {
+	testMode := opts.Globals.TestMode
+	result := make(abi.AbiInterfaceMap)
+	if opts.Known {
+		if err = abi.PreloadKnownAbis(opts.Globals.Chain, result, false); err != nil {
+			return
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	fetchData := func(modelChan chan types.Modeler[types.RawFunction], errorChan chan error) {
+		// result := make(abi.AbiInterfaceMap)
+		for _, addr := range opts.Addrs {
+			address := common.HexToAddress(addr)
+			err = abi.LoadAbiFromAddress(opts.Globals.Chain, address, result)
+			if err == nil {
+				if testMode {
+					// Tests expect sorted map, because maps in C++ are sorted.
+					keys := make([]string, 0, len(result))
+
+					for k := range result {
+						keys = append(keys, k)
+					}
+					sort.Strings(keys)
+
+					for _, k := range keys {
+						modelChan <- result[k]
+					}
+					continue
+				}
+				for _, function := range result {
+					modelChan <- function
+				}
+				continue
+			}
+			if !os.IsNotExist(err) {
+				errorChan <- err
+				cancel()
+			}
+
+			// We didn't find the file
+			// Check if the address is a contract
+			contract, err := account.IsContractAt(
+				opts.Globals.Chain,
+				address,
+				nil, // use latest block number
+			)
+			if err != nil {
+				errorChan <- err
+				cancel()
+			}
+			if !contract {
```

Comment:
Couldn't we have done this in the Validate function so we don't even get this far if the user includes non-smart contracts to the tool? (I realize that this is probably how it works in the C++ code, but should it?)

Reply:
We could, but my understanding of C++ code is that it ignores non-contract addresses, but continues. Putting this into Validate would change this behaviour to "I will not run unless all of the addresses are contracts". We'd also need another loop to check all addresses before running the actual code. But I can be mistaken